README.md
<p align="center">
    <img src="https://user-images.githubusercontent.com/56393921/130352761-91b39f46-5f1d-4131-a8c0-d118bf77c48f.png">
</p>
<p align="center">
  
[![PyPI version](https://badge.fury.io/py/usgslidar.svg)](https://badge.fury.io/py/usgslidar) 
![GitHub issues](https://img.shields.io/github/issues-raw/Azariagmt/AgriTech-USGS-LIDAR)
![GitHub](https://img.shields.io/github/license/Azariagmt/AgriTech-USGS-LIDAR)
[![Documentation Status](https://readthedocs.org/projects/usgslidar/badge/?version=latest)](https://usgslidar.readthedocs.io/en/latest/?badge=latest)
 </p>
 
<div>  
  The USGS recently released high resolution elevation data as a lidar point cloud called <a href="https://www.usgs.gov/core-science-systems/ngp/3dep">USGS 3DEP</a> in a <a href="https://registry.opendata.aws/usgs-lidar/">public dataset on Amazon</a>. This dataset is essential to build models of water flow and predict plant health and maize harvest. 
</div>

<br>
<div>
  The purpose of this package (<b><a href="https://pypi.org/project/usgslidar/">usgslidar</a></b>) is to produce an easy to use, reliable and well designed python module that domain experts and data scientists can use to fetch, visualise, and transform publicly available satellite and LIDAR data. In particular, it interfaces with USGS 3DEP and fetches data available in the s3 instance. 
</div>

<h2>Table of content</h2>

1. <a href="#setup">Setup</a>
2. <a href="#usage">Usage</a>
3. <a href="#starting">Getting started (tutorial)</a>
4. <a href="#example">Example notebook</a>

<h3 id="setup"> Setup </h3>
To build our pipeline we actually use <a href="https://pdal.io/tutorial/iowa-entwine.html">PDAL</a> under the hood. Since PDAL doesnt build properly with pip in the current version, you need to use conda to manage your dependencies. 
<br>

<p>

  * Install [Anaconda](https://www.anaconda.com/)
  * In a conda environment install PDAL and python-pdal
    ```bash
    conda install -c conda-forge pdal python-pdal
    ```
  * Install usgslidar
    ```bash
    pip install usgslidar
    ```
</p>

<h3 id="usage"> Usage </h3>
Full Documentation for usgslidar package can be found <a href="https://usgslidar.readthedocs.io/en/latest/?">here</a>.
<h3 id="starting"> Getting started </h3>
Lets view an example of visualizing a point cloud data on a farm in Iowa
<p>

  * From the get_data module lets import the get_geopandas_dataframe module
      ```python
      from usgslidar.get_data import get_geopandas_dataframe
      ```
    
  * We will get back the elevation and geometry dataframe when we call the get_geopandas_dataframe function
    ```python
    elevation_df = get_geopandas_dataframe()
    ```
    
  * Lets now import the plot_3d_map function from the visualize module to help us plot the 3d map
    ```python
    from usgslidar.visualize import plot_3d_map
    ```
    
  * We can now plot the 3d map calling the function and passing in the elevation dataframe we got above
    ```python
    plot_3d_map(df=elevation_df)
    ```
    
  * We will get an output of a 3d plot like below that we can play around with
    
  ![output](https://user-images.githubusercontent.com/56393921/130657829-826d2fe0-3857-4da6-a722-2ccd2875f985.png)
  
    
</p>

<h3 id="example"> Examples </h3>
Example notebooks with specific purposes can be found <a href="https://github.com/Azariagmt/usgslidar/tree/main/demo-notebooks">here.</a>


CHANGELOG.txt
Change Log 

===============
0.0.17
    fixes logs directory and import error when running get_data script
0.0.18
===============
0.0.19
===============
0.0.20
===============
0.0.21
===============
0.0.22
===============
0.0.23
    Fixes FileNotFoundError: [Errno 2] No such file or directory: 'get_data.json'
        json PDAL pipeline wasnt getting packaged properly this version fixes that
===============
0.0.24
    Fixes log dir not being found if it has been created once
0.0.25
    Fixes json path not being found within package
    fixes logger and module error on get_data module

usgslidar/logs.py
import os
import logging
# a function  to create and save logs in the log files


def log(path, file):
    """[Create a log file to record the experiment's logs]
    Arguments:
        path {string} -- path to the directory
        file {string} -- file name
    Returns:
        [obj] -- [logger that record logs]
    """

    # check if the file exist
    log_file = os.path.join(path, file)

    if not os.path.isfile(log_file):
        open(log_file, "w+").close()

    console_logging_format = "%(levelname)s %(message)s"
    file_logging_format = "%(levelname)s: %(asctime)s: %(message)s"

    # configure logger
    logging.basicConfig(level=logging.INFO, format=console_logging_format)
    logger = logging.getLogger()

    # create a file handler for output file
    handler = logging.FileHandler(log_file)

    # set the logging level for log file
    handler.setLevel(logging.INFO)

    # create a logging format
    formatter = logging.Formatter(file_logging_format)
    handler.setFormatter(formatter)

    # add the handlers to the logger
    logger.addHandler(handler)

    return logger

usgslidar/visualize.py
import numpy as np
import laspy as lp
import matplotlib.pyplot as plt
from mpl_toolkits import mplot3d
# from get_data import get_geopandas_dataframe
import geopandas as gpd
from .logs import log
import os

if (not os.path.isdir('./logs')):
    os.mkdir("./logs")

logs_path = "./logs/visualize.logs"
if not os.path.exists(logs_path):
    with open(logs_path, "w"):
        global logger
        logger = log(path="./logs/", file="visualize.logs")
        logger.info("Starts Visualize script")
else:
    logger = log(path="./logs/", file="visualize.logs")
    logger.info("Starts Visualize script")


input_path = "../data/laz/"
dataname = "SoPlatteRiver"

point_cloud = lp.read(input_path+dataname+".las")

colors = np.vstack((point_cloud.red, point_cloud.green,
                   point_cloud.blue)).transpose()


def plot_3d_map(df=None):
    """Plots 3d map of input dataframe

    Arguments: df, takes in dataframe 
    return: None

    """
    if df['elevation'][0] != None:
        gdf = df
        gdf.crs = "epsg:4326"
        print(gdf)

    x = gdf.geometry.x
    y = gdf.geometry.y
    z = gdf.elevation
    points = np.vstack((x, y, z)).transpose()
    factor = 160
    decimated_points = points[::]
    decimated_colors = colors[::]
    print(len(decimated_colors))
    print(decimated_colors)
    fig, ax = plt.subplots(1, 1, figsize=(12, 10))
    ax = plt.axes(projection='3d')
    ax.scatter(decimated_points[:, 0], decimated_points[:, 1],
               decimated_points[:, 2],  s=0.01, color=decimated_colors/255)
    plt.show()


usgslidar/get_data.py
import pdal
import json
import geopandas as gpd
from pyproj import Transformer
from .logs import log
from shapely.geometry import Polygon, Point
import os

if (not os.path.isdir('./logs')):
    os.mkdir("./logs")

logs_path = "./logs/get_data.logs"
if not os.path.exists(logs_path):
    with open(logs_path, "w"):
        global logger
        logger = log(path="./logs/", file="get_data.logs")
        logger.info("Starts Get data script")
else:
    logger = log(path="./logs/", file="get_data.logs")
    logger.info("Starts Get data script")

PUBLIC_DATA_PATH = "https://s3-us-west-2.amazonaws.com/usgs-lidar-public/"

REGION = "IA_FullState/"

MINX, MINY, MAXX, MAXY = [-93.756155, 41.918015, -93.747334, 41.921429]
BOUNDS = f"({[MINX,MAXX]},{[MINY,MAXY]})"
BOUNDS = "([-11669524, -11666600.8],[4776607.3, 4778714.4])"

PUBLIC_ACCESS_PATH = PUBLIC_DATA_PATH+REGION+"ept.json"

# TODO ask user to input filename
OUTPUT_FILENAME_LAZ = "../data/laz/SoPlatteRiver.las"
OUTPUT_FILENAME_TIF = "../data/tif/SoPlatteRiver.tif"


here = os.path.dirname(os.path.abspath(__file__))
PIPELINE_PATH = os.path.join(here, "get_data.json")

def get_polygon_boundaries(polygon: Polygon, input_epsg: int = 3857, output_epsg: int = 4326):
    """Gets the required polygon boundaries

    Arguments:
        Polygon -- A shapely polygon
                   https://shapely.readthedocs.io/en/stable/manual.html
        Input epsg -- The input CRS reference system
                      default -- 3857
        Output epsg -- The output CRS reference system
                      default -- 26915

    Returns: A tuple of lists containing rectangular boundary
            ([minx, maxx],[miny,maxy]) &
            the initial user polygon input

    """
    polygon_df = gpd.GeoDataFrame([polygon], columns=['geometry'])

    polygon_df.set_crs(epsg=output_epsg, inplace=True)
    polygon_df['geometry'] = polygon_df['geometry'].to_crs(
        epsg=input_epsg)
    minx, miny, maxx, maxy = polygon_df['geometry'][0].bounds

    polygon_input = 'POLYGON(('

    xcord, ycord = polygon_df['geometry'][0].exterior.coords.xy
    for x, y in zip(list(xcord), list(ycord)):
        polygon_input += f'{x} {y}, '
    polygon_input = polygon_input[:-2]
    polygon_input += '))'

    print(polygon_input)

    return f"({[minx, maxx]},{[miny,maxy]})", polygon_input


def get_raster_terrain(polygon: list, region: str, PUBLIC_ACCESS_PATH: str = PUBLIC_ACCESS_PATH,
                       OUTPUT_FILENAME_LAZ: str = OUTPUT_FILENAME_LAZ, OUTPUT_FILENAME_TIF: str = OUTPUT_FILENAME_TIF,
                       PIPELINE_PATH: str = PIPELINE_PATH
                       ) -> None:
    """Initializes the PDAL pipeline and gets bound for entire region

    Arguments: 
        Polygon (list of x,y coordinate tuples) -- sets boundary
        region (string) -- region

    Optional Arguments:
        PIPELINE_PATH (string) -- path to new PDAL pipeline

    Returns:
        Arrays from the PDAL pipeline
    """
    try:
        input_polygon = Polygon(polygon)
        MINX, MINY, MAXX, MAXY = [-93.756155, 41.918015, -93.747334, 41.921429]
        input_polygon = Polygon(
            ((MINX, MINY), (MINX, MAXY), (MAXX, MAXY), (MAXX, MINY), (MINX, MINY)))
        logger.info("Polygon input success!")
    except RuntimeError as e:
        logger.exception("Polygon input error")

    try:
        bounds, polygon_input = get_polygon_boundaries(input_polygon)
        logger.info("Rectangular input extracted successfully")
    except RuntimeError as e:
        logger.exception("Polygon extraction error")

    # TODO convert CRS to 3857

    try:
        with open(PIPELINE_PATH) as json_file:
            the_json = json.load(json_file)

        the_json['pipeline'][0]['bounds'] = bounds
        the_json['pipeline'][0]['filename'] = PUBLIC_ACCESS_PATH

        the_json['pipeline'][1]['polygon'] = polygon_input

        # TODO replace 4326 with out_srs
        # the_json['pipeline'][4]['out_srs'] = "EPSG:4326"

        the_json['pipeline'][4]['filename'] = OUTPUT_FILENAME_LAZ
        # the_json['pipeline'][6]['filename'] = OUTPUT_FILENAME_TIF

        pipeline = pdal.Pipeline(json.dumps(the_json))

        logger.info("pipeline initiated")
    except RuntimeError as e:
        logger.exception("failed to initiate pipeline")
    try:
        pipe_exec = pipeline.execute()
        print(pipe_exec)
        # TODO all the comments
        # list df numpy arrays...x, y and z bounds
        # laz file contains a lot of info
        # interested in elevation point
        # z is the elevation
        metadata = pipeline.metadata
        print(metadata)
        logger.info("Pipeline executed and metadata generated successfully")
    except RuntimeError as e:
        logger.exception("Failed to generate metadata and execute pipeline")

    try:
        print("PIPE EXEC return")
        pipeline_arrays = pipeline.arrays
        print(pipeline_arrays)
        logger.info("Pipeline arrays successfully returned.")
        return pipeline_arrays
    except RuntimeError as e:
        logger.exception("Failed to retrieve pipeline arrays")


def get_elevetion(array_data, crs_epgs=4326):
    """Returns the elevation of each point in the point_cloud data

    """
    if array_data:
        for i in array_data:
            geometry_points = [Point(x, y) for x, y in zip(i["X"], i["Y"])]
            elevetions = i["Z"]
            df = gpd.GeoDataFrame(columns=["elevation", "geometry"])
            df['elevation'] = elevetions
            df['geometry'] = geometry_points
            df = df.set_geometry("geometry")
            df.set_crs(epsg=crs_epgs, inplace=True)
        print(df)
        df.to_file("../data/elevation.geojson", driver='GeoJSON')
        return df

    return None


def get_geopandas_dataframe():
    """Returns the Geopandas Dataframe


    """
    pipeline_arrays = get_raster_terrain(polygon=[
                                         (-93.756155, 41.918015), (-93.747334, 41.921429), (-93.756155, 41.918015), (-93.756155, 41.918015)], region="IA_FullState")
    elevation_df = get_elevetion(pipeline_arrays)
    return elevation_df


if __name__ == "__main__":
    parrays = get_raster_terrain(polygon=[(-93.756155, 41.918015), (-93.747334, 41.921429),
                                 (-93.756155, 41.918015), (-93.756155, 41.918015)], region="IA_FullState")
    get_elevetion(parrays)
